File: typescript/tests/integration/mobile-agent-streaming.integration.test.ts
/**
 * Integration tests for Mobile Agent streaming output feature.
 *
 * Tests the WS-based streaming execution path with real backend services.
 * Uses debug image imgc-0ab5takhnmlvhx9gp for Mobile Agent.
 *
 * Note: These tests create real sessions and should not be run concurrently
 * to avoid rate limiting and resource exhaustion.
 */

import { AgentBay } from '../../src/agent-bay';
import { AgentEvent, AgentStreamingOptions } from '../../src/agent/agent';
import { Session } from '../../src/session';
import { log } from '../../src/utils/logger';

describe('Mobile Agent Streaming Integration Tests', () => {
  let agentBay: AgentBay;
  let session: Session;
  const apiKey = process.env.AGENTBAY_API_KEY;

  beforeAll(async () => {
    if (!apiKey) {
      console.warn('Warning: AGENTBAY_API_KEY not set. Tests will be skipped.');
      return;
    }

    agentBay = new AgentBay({ apiKey });
    await new Promise((r) => setTimeout(r, 3000));

    const sessionResult = await agentBay.create({
      imageId: 'imgc-0ab5takhnmlvhx9gp',
    });

    if (!sessionResult.success || !sessionResult.session) {
      console.error(`Failed to create session: ${sessionResult.errorMessage}`);
      return;
    }

    session = sessionResult.session;
    log(`Session created: ${session.sessionId}`);
  });

  afterAll(async () => {
    if (session) {
      try {
        log(`Cleaning up session: ${session.sessionId}`);
        await agentBay.delete(session);
        log(`Session deleted: ${session.sessionId}`);
      } catch (e) {
        log(`Error deleting session: ${e}`);
      }
    }
  });

  test('should stream events with typed callbacks', async () => {
    if (!session) {
      log('Skipping: session not created');
      return;
    }

    const reasoningEvents: AgentEvent[] = [];
    const contentEvents: AgentEvent[] = [];
    const toolCalls: AgentEvent[] = [];
    const toolResults: AgentEvent[] = [];

    const options: AgentStreamingOptions = {
      onReasoning: (event) => {
        reasoningEvents.push(event);
        log(`[Reasoning] round=${event.round}: ${(event.content || '').substring(0, 100)}...`);
      },
      onContent: (event) => {
        contentEvents.push(event);
        log(`[Content] round=${event.round}: ${(event.content || '').substring(0, 100)}...`);
      },
      onToolCall: (event) => {
        toolCalls.push(event);
        log(`[ToolCall] round=${event.round}: ${event.toolName}(${JSON.stringify(event.args)})`);
      },
      onToolResult: (event) => {
        toolResults.push(event);
        log(`[ToolResult] round=${event.round}: ${event.toolName} -> ${JSON.stringify(event.result).substring(0, 100)}...`);
      },
    };

    log('Testing mobile agent streaming with typed callbacks');

    const result = await session.agent.mobile.executeTaskAndWait(
      'Open Settings app',
      180,
      10,
      options,
    );

    log(`Result: success=${result.success}, status=${result.taskStatus}`);
    log(`Reasoning: ${reasoningEvents.length}, Content: ${contentEvents.length}, ToolCalls: ${toolCalls.length}, ToolResults: ${toolResults.length}`);

    expect(typeof result.taskStatus).toBe('string');
  }, 300000);
});

File: golang/tests/pkg/integration/mobile_agent_streaming_integration_test.go
package integration_test

import (
	"os"
	"testing"
	"time"

	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay"
	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay/agent"
)

func TestMobileAgentStreaming_TypedCallbacks(t *testing.T) {
	apiKey := os.Getenv("AGENTBAY_API_KEY")
	if apiKey == "" {
		t.Skip("AGENTBAY_API_KEY not set")
	}

	client, err := agentbay.NewAgentBay(apiKey, nil)
	if err != nil {
		t.Fatalf("Failed to create AgentBay client: %v", err)
	}

	time.Sleep(3 * time.Second)

	params := agentbay.NewCreateSessionParams().WithImageId("imgc-0ab5takhnmlvhx9gp")
	sessionResult, err := client.Create(params)
	if err != nil {
		t.Fatalf("Failed to create session: %v", err)
	}
	if sessionResult.Session == nil {
		t.Fatal("Session is nil")
	}
	defer func() {
		sessionResult.Session.Delete()
	}()

	t.Logf("Session created: %s", sessionResult.Session.SessionID)

	var reasoningCount, contentCount, toolCallCount, toolResultCount int

	result := sessionResult.Session.Agent.Mobile.ExecuteTaskAndWait(
		"Open Settings app",
		180,
		agent.MobileTaskOptions{
			MaxSteps: 10,
			StreamOptions: agent.StreamOptions{
				OnReasoning: func(evt agent.AgentEvent) {
					reasoningCount++
					t.Logf("[Reasoning] round=%d: %s", evt.Round, truncate(evt.Content, 100))
				},
				OnContent: func(evt agent.AgentEvent) {
					contentCount++
					t.Logf("[Content] round=%d: %s", evt.Round, truncate(evt.Content, 100))
				},
				OnToolCall: func(evt agent.AgentEvent) {
					toolCallCount++
					t.Logf("[ToolCall] round=%d: %s", evt.Round, evt.ToolName)
				},
				OnToolResult: func(evt agent.AgentEvent) {
					toolResultCount++
					t.Logf("[ToolResult] round=%d: %s", evt.Round, evt.ToolName)
				},
			},
		},
	)

	t.Logf("Result: success=%v, status=%s", result.Success, result.TaskStatus)
	t.Logf("Reasoning: %d, Content: %d, ToolCalls: %d, ToolResults: %d",
		reasoningCount, contentCount, toolCallCount, toolResultCount)
}

func truncate(s string, maxLen int) string {
	if len(s) <= maxLen {
		return s
	}
	return s[:maxLen] + "..."
}
